src/lib/efthrm/tcp_helper_linux.c

```diff
@@ -60,15 +68,83 @@
 
 #define FOP_RW_ITER_CALL_BASE_HANDLER(base_handler, iocb, iter)         \
   do {                                                                  \
-    return base_handler(iocb->ki_filp, iter_iov(v), v->nr_segs);        \
+    return base_handler(iocb->ki_filp, iter_iov(v), v->nr_segs, CI_ADDR_SPC_CURRENT); \
   } while (0);
 
 #endif /* EFRM_HAVE_ITER_UBUF */
 
+#ifdef EFRM_HAVE_FOP_SENDPAGE
+#define oo_fop_splice_write generic_splice_sendpage
+#define oo_fop_splice_read generic_file_splice_read
+
+static ssize_t rw_iter_pipe_handler(fop_rw_base_handler base_handler,
+                                    struct kiocb *iocb, struct iov_iter *v)
+{
+  struct iovec iov;
+  size_t len = iov_iter_count(v);
+  ssize_t rc;
+
+  iov.iov_base = kzalloc(len, GFP_KERNEL);
+  iov.iov_len = len;
+
+  rc = base_handler(iocb->ki_filp, &iov, 1, CI_ADDR_SPC_KERNEL);
+  /* This handler is called only in splice_read() case, i.e. from
+   * socket to pipe. We don't have to care about another direction,
+   * copy_to_iter() is enough. */
+  if( rc > 0 )
+    rc = copy_to_iter(iov.iov_base, iov.iov_len, v);
```

Comment:
Could you clarify why we pass `iov.iov_len` and not `rc` to `copy_to_iter()`? AFAIU, `iov.iov_len == iov_iter_count()` is the capacity, but `rc` is the number of bytes actually "retrieved" from the base handler such as `linux_tcp_helper_fop_read_iov_pipe()`, so it seems we don't want to exceed `rc`.

Reply:
Yes, you're right! `rc` must be there. Thanks.
I fixed that.

Reply:
Excellent, thank you!